### DISCUSSION TITLE
Regexp engine to use negative lookahead assertion

### DISCUSSION THREAD

**Discussion by h3ct0rjs** [BurntSushi/ripgrep#2737] [Q&A] [answered]
Hello Guys, hope you're doing great today. I'm currently building a couple of things that require that I use regexp. I'm amazed to use rg in my development, however, I'm not able to use my regexp. I'm currently using a negative lookahead assertion in my regexp but it seems like RG is not able to understand this. 

as a general example I have the following regexp : 

`\b\w+-company\b(?![\w-])`

This regular expression will match "username-company" but will ignore occurrences where "company" is followed by any other characters.

I'm currently trying to use rg as follows : 

` rg -e "\b\w+-company\b(?![\w-])" . --json`

Is there any option to use my [negative lookahead assertion ](https://www.regular-expressions.info/lookaround.html)in my regexp. 

Regards, 
H

**Comment by BurntSushi** [comment, marked as answer]
From the README:

> ripgrep has optional support for switching its regex engine to use PCRE2. Among other things, this makes it possible to use look-around and backreferences in your patterns, which are not supported in ripgrep's default regex engine. PCRE2 support can be enabled with -P/--pcre2 (use PCRE2 always) or --auto-hybrid-regex (use PCRE2 only if needed). An alternative syntax is provided via the --engine (default|pcre2|auto-hybrid) option.

ripgrep uses [Rust's regex crate](https://docs.rs/regex/latest/regex/) by default. From its docs:

> This crate provides routines for searching strings for matches of a [regular expression](https://en.wikipedia.org/wiki/Regular_expression) (aka “regex”). The regex syntax supported by this crate is similar to other regex engines, but it lacks several features that are not known how to implement efficiently. This includes, but is not limited to, look-around and backreferences.

**Reply by h3ct0rjs** [reply]
@BurntSushi thank you very much for your answer. I will be running my test 😄 again. 
Just curious, is there any specific reason to not use PCRE2 as default in the tool ? or it is done to be more compatible with the current shell tools ?

**Reply by BurntSushi** [reply]
Because the default is faster for grep workloads (generally speaking) and doesn't suffer from catastrophic backtracking like PCRE2 does. PCRE2 support is an optional feature of ripgrep to appease folks who want fancier regex features. ripgrep did not originally have PCRE2 support.

Also, I wrote ripgrep's default regex engine. ripgrep was born originally to help me benchmark the regex engine.

See also (ripgrep uses `rust/regex`): https://github.com/BurntSushi/rebar

**Reply by h3ct0rjs** [reply]
Thank @BurntSushi, keep up the great work!
